File: library/kotlin/io/envoyproxy/envoymobile/StreamCallbacks.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.types.EnvoyHTTPCallbacks
import io.envoyproxy.envoymobile.engine.types.EnvoyStreamIntel
import io.envoyproxy.envoymobile.engine.types.EnvoyFinalStreamIntel
import java.nio.ByteBuffer
import java.util.concurrent.Executor

/**
 * A collection of platform-level callbacks that are specified by consumers
 * who wish to interact with streams.
 *
 * `StreamCallbacks` are bridged through to `EnvoyHTTPCallbacks` to communicate with the engine.
 */
internal class StreamCallbacks {
  var onHeaders: (
    (headers: ResponseHeaders, endStream: Boolean, streamIntel: StreamIntel) -> Unit
  )? = null
  var onData: ((data: ByteBuffer, endStream: Boolean, streamIntel: StreamIntel) -> Unit)? = null
  var onTrailers: ((trailers: ResponseTrailers, streamIntel: StreamIntel) -> Unit)? = null
  var onCancel: ((streamIntel: StreamIntel, finalStreamIntel: FinalStreamIntel) -> Unit)? = null
  var onError: ((error: EnvoyError, streamIntel: StreamIntel,
                 finalStreamIntel: FinalStreamIntel) -> Unit)? = null
  var onSendWindowAvailable: ((streamIntel: StreamIntel) -> Unit)? = null
  var onComplete: ((streamIntel: StreamIntel, finalStreamIntel: FinalStreamIntel) -> Unit)? = null
}

/**
 * Class responsible for bridging between the platform-level `StreamCallbacks` and the
 * engine's `EnvoyHTTPCallbacks`.
 */
internal class EnvoyHTTPCallbacksAdapter(
  private val executor: Executor,
  private val callbacks: StreamCallbacks
) : EnvoyHTTPCallbacks {
  override fun getExecutor(): Executor {
    return executor
  }

  override fun onHeaders(
    headers: Map<String, List<String>>,
    endStream: Boolean,
    streamIntel: EnvoyStreamIntel
  ) {
    callbacks.onHeaders?.invoke(ResponseHeaders(headers), endStream, StreamIntel(streamIntel))
  }

  override fun onData(byteBuffer: ByteBuffer, endStream: Boolean, streamIntel: EnvoyStreamIntel) {
    callbacks.onData?.invoke(byteBuffer, endStream, StreamIntel(streamIntel))
  }

  override fun onTrailers(trailers: Map<String, List<String>>, streamIntel: EnvoyStreamIntel) {
    callbacks.onTrailers?.invoke(ResponseTrailers((trailers)), StreamIntel(streamIntel))
  }

  override fun onError(
    errorCode: Int,
    message: String,
    attemptCount: Int,
    streamIntel: EnvoyStreamIntel,
    finalStreamIntel: EnvoyFinalStreamIntel
  ) {
    callbacks.onError?.invoke(
      EnvoyError(errorCode, message, attemptCount),
      StreamIntel(streamIntel),
      FinalStreamIntel(finalStreamIntel)
    )
  }

  override fun onCancel(streamIntel: EnvoyStreamIntel, finalStreamIntel: EnvoyFinalStreamIntel) {
    callbacks.onCancel?.invoke(StreamIntel(streamIntel), FinalStreamIntel(finalStreamIntel))
  }

  override fun onSendWindowAvailable(streamIntel: EnvoyStreamIntel) {
    callbacks.onSendWindowAvailable?.invoke(StreamIntel(streamIntel))
  }

  override fun onComplete(streamIntel: EnvoyStreamIntel, finalStreamIntel: EnvoyFinalStreamIntel) {
    callbacks.onComplete?.invoke(StreamIntel(streamIntel), FinalStreamIntel(finalStreamIntel))
  }
}

File: library/swift/StreamCallbacks.swift
import Dispatch
@_implementationOnly import EnvoyEngine
import Foundation

/// A collection of platform-level callbacks that are specified by consumers
/// who wish to interact with streams.
///
/// `StreamCallbacks` are bridged through to `EnvoyHTTPCallbacks` to communicate with the engine.
final class StreamCallbacks {
  var onHeaders: (
    (_ headers: ResponseHeaders, _ endStream: Bool, _ streamIntel: StreamIntel) -> Void
  )?
  var onData: ((_ body: Data, _ endStream: Bool, _ streamIntel: StreamIntel) -> Void)?
  var onTrailers: ((_ trailers: ResponseTrailers, _ streamIntel: StreamIntel) -> Void)?
  var onComplete: ((_ streamIntel: StreamIntel, _ finalStreamintel: FinalStreamIntel) -> Void)?
  var onCancel: ((_ streamIntel: StreamIntel, _ finalStreamintel: FinalStreamIntel) -> Void)?
  var onError: ((_ error: EnvoyError, _ streamIntel: StreamIntel, _ finalStreamIntel: FinalStreamIntel) -> Void)?
}

extension EnvoyHTTPCallbacks {
  /// Initializer propagating the platform callbacks into callbacks that the engine can use.
  ///
  /// - parameter callbacks: Platform callbacks to use.
  /// - parameter queue:     Queue on which to receive callback events.
  convenience init(callbacks: StreamCallbacks, queue: DispatchQueue) {
    self.init()
    self.dispatchQueue = queue
    self.onHeaders = { callbacks.onHeaders?(ResponseHeaders(headers: $0), $1, StreamIntel($2)) }
    self.onData = { callbacks.onData?($0, $1, StreamIntel($2)) }
    self.onTrailers = { callbacks.onTrailers?(ResponseTrailers(headers: $0), StreamIntel($1)) }
    self.onComplete = { callbacks.onCancel?(StreamIntel($0), FinalStreamIntel($1)) }
    self.onCancel = { callbacks.onCancel?(StreamIntel($0), FinalStreamIntel($1)) }
    self.onError = { errorCode, message, attemptCount, streamIntel, finalStreamIntel in
      // The initializer below will return nil if `attemptCount` is negative.
      // This is the desired behavior because the bridge layer uses -1 to signify absence.
      let error = EnvoyError(errorCode: errorCode, message: message,
                             attemptCount: UInt32(exactly: attemptCount), cause: nil)
      callbacks.onError?(error, StreamIntel(streamIntel), FinalStreamIntel(finalStreamIntel))
    }
  }
}
